### PR TITLE
debugserver: Add /healthz and /ready endpoints

### PR DESCRIPTION
**Pull request by efritz**
This PR adds /healthz and /ready endpoints to the debug server, which is generally started at the top of the application's main function.

This wasn't true for a few services (repo-updater and the frontend), which is corrected here.

Debug servers will respond to 200 OK to all /healthz requests, and will conditionally respond to 200 OK to /ready requests when the channel passed to the debug server is closed. When to close this is on an app-by-app basis. We used the following determination for each service:

- If it does not touch the database, it's ready immediately
- If it touches the database, we wait until the database has migrated before becoming ready

Fixes https://github.com/sourcegraph/sourcegraph/issues/19757 (will need updates on compose and k8s to respect ready status).

### CODE REVIEW

**Review by davejrt — APPROVED** [review]

**Comment by unknwon** [comment]
> There's a big diff but its' all due to indentation level

CleanShot 2021-04-09 at 03 53 51@2x

**Comment by efritz** [comment]
image

File: internal/debugserver/ready.go [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+// readyHandler returns an http.HandlerFunc that responds to the /ready
+// requests on the debugserver port. This will return a 200 OK once the
+// given channel is closed, and a 500 Internal Server Error otherwise.
+func readyHandler(ready <-chan struct{}) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		select {
+		case <-ready:
+			w.WriteHeader(http.StatusOK)
+		default:
+			w.WriteHeader(http.StatusInternalServerError)
```
Comment (unknwon) [inline internal/debugserver/ready.go:21]: Would 503 (service not available) be more appropriate?

File: cmd/repo-updater/shared/main.go [hunk 2]
```diff
@@ -253,96 +287,99 @@ func Main(enterpriseInit EnterpriseInit) {
...
+			http.Error(w, "failed to marshal infos: "+err.Error(), http.StatusInternalServerError)
+			return
+		}
+		w.Header().Set("Content-Type", "application/json")
+		_, _ = w.Write(resp)
+	})
+
+	// Migrations may take a while, but after they're done we'll immediately
+	// spin up a server and can accept traffic. Inform external clients we'll
+	// be ready for traffic.
```
Comment (unknwon) [inline cmd/repo-updater/shared/main.go:377]: I totally understand what this block of comment is saying, but having hard time finding where we wait for migrations to be done?

File: internal/debugserver/ready.go (hunk 1 above)
Comment (efritz) [reply internal/debugserver/ready.go]: Changed.

File: cmd/repo-updater/shared/main.go (hunk 2 above)
Comment (efritz) [reply cmd/repo-updater/shared/main.go:377]: Updated comments to clarify.

**Review by unknwon — APPROVED** [review]
Nice st~~a~~uff!

**Pull request merged**